File: platform/aws/provider/deploy/assets.go
package deploy

import (
	"context"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"golang.org/x/sync/errgroup"

	"github.com/ocelhq/ocel/pkg/naming"
	"github.com/ocelhq/ocel/pkg/providerkit"
)

const staticAssetsDir = "static"

const (
	immutableCacheControl  = "public, max-age=31536000, immutable"
	revalidateCacheControl = "public, max-age=0, must-revalidate"
)

const (
	nextStaticSegment    = "_next/static/"
	serviceWorkerSegment = "service-worker/"
)

var assetContentTypes = map[string]string{
	".html":        "text/html; charset=utf-8",
	".js":          "text/javascript; charset=utf-8",
	".mjs":         "text/javascript; charset=utf-8",
	".css":         "text/css; charset=utf-8",
	".json":        "application/json; charset=utf-8",
	".map":         "application/json; charset=utf-8",
	".svg":         "image/svg+xml",
	".png":         "image/png",
	".jpg":         "image/jpeg",
	".jpeg":        "image/jpeg",
	".gif":         "image/gif",
	".webp":        "image/webp",
	".avif":        "image/avif",
	".ico":         "image/x-icon",
	".woff":        "font/woff",
	".woff2":       "font/woff2",
	".ttf":         "font/ttf",
	".eot":         "application/vnd.ms-fontobject",
	".txt":         "text/plain; charset=utf-8",
	".xml":         "application/xml",
	".webmanifest": "application/manifest+json",
	".wasm":        "application/wasm",
}

var metadataContentTypes = map[string]string{
	"robots.txt":    "text/plain",
	"manifest.json": "application/manifest+json",
}

func assetContentType(rel string) string {
	name := strings.ToLower(rel[strings.LastIndex(rel, "/")+1:])
	if ct, ok := metadataContentTypes[name]; ok {
		return ct
	}
	dot := strings.LastIndex(name, ".")
	if dot == -1 {
		return "application/octet-stream"
	}
	if ct, ok := assetContentTypes[name[dot:]]; ok {
		return ct
	}
	return "application/octet-stream"
}

func assetCacheControl(rel string) string {
	path := "/" + rel
	at := strings.Index(path, "/"+nextStaticSegment)
	if at == -1 {
		return revalidateCacheControl
	}
	item := path[at+len("/"+nextStaticSegment):]
	if strings.HasPrefix(item, serviceWorkerSegment) {
		return revalidateCacheControl
	}
	return immutableCacheControl
}

func assetHeaders(rel string) objectHeaders {
	return objectHeaders{contentType: assetContentType(rel), cacheControl: assetCacheControl(rel)}
}

const imageConfigFile = "image-config.json"

func appAssetPrefix(c naming.Coordinate) string {
	return c.AssetKey("")
}

func assetPlaneTargets(cfg Config) []uploadTarget {
	return []uploadTarget{
		{up: cfg.CacheStoreUploader, bucket: cfg.CacheStoreBucket, class: cfg.Class},
		{up: cfg.Uploader, bucket: cfg.AssetBucket, class: cfg.Class},
	}
}

type assetUpload struct {
	key, src string
	to       []uploadTarget
	replace  bool
	headers  objectHeaders
}

func staticAssetSet(cfg Config, app, runtime string, coord naming.Coordinate) (*assetSet, error) {
	if runtime != runtimeNext {
		return nil, nil
	}
	if cfg.CacheStoreBucket == "" || cfg.CacheStoreUploader == nil {
		return nil, nil
	}

	assetBucket := uploadTarget{up: cfg.Uploader, bucket: cfg.AssetBucket, class: cfg.Class}
	plane := assetPlaneTargets(cfg)
	var uploads []assetUpload
	manifest := newSetManifest()
	root := appArtifactRoot(cfg.ArtifactRoot, app)
	dir := filepath.Join(root, staticAssetsDir)
	files, err := collectFiles(dir)
	if err != nil {
		return nil, err
	}
	for _, file := range files {
		key := coord.AssetKey(file.rel)
		uploads = append(uploads, assetUpload{
			key:     key,
			src:     filepath.Join(dir, filepath.FromSlash(file.rel)),
			to:      plane,
			headers: assetHeaders(file.rel),
		})
		for _, to := range plane {
			manifest.add(to.bucket, key, file.size)
		}
	}
	imageConfig := filepath.Join(root, imageConfigFile)
	switch info, err := os.Stat(imageConfig); {
	case err == nil:
		uploads = append(uploads, assetUpload{
			key:     coord.ImageConfigKey(),
			src:     imageConfig,
			to:      []uploadTarget{assetBucket},
			replace: true,
			headers: objectHeaders{contentType: "application/json"},
		})
		manifest.add(assetBucket.bucket, coord.ImageConfigKey(), info.Size())
	case !errors.Is(err, fs.ErrNotExist):
		return nil, fmt.Errorf("stat image config for %s: %w", app, err)
	}
	if len(uploads) == 0 {
		return nil, nil
	}
	if err := assetBucket.validate(); err != nil {
		return nil, err
	}

	return &assetSet{
		name:   staticAssetSetName,
		app:    app,
		files:  manifest.files,
		digest: manifest.digest(),
		push: func(ctx context.Context, report providerkit.Reporter) error {
			return pushStaticAssets(ctx, app, uploads, report)
		},
	}, nil
}

func pushStaticAssets(ctx context.Context, app string, uploads []assetUpload, report providerkit.Reporter) error {
	say(report, "Uploading "+app+"'s static assets")
	phaseStart := time.Now()
	g, ctx := errgroup.WithContext(ctx)
	g.SetLimit(uploadConcurrency)
	stats := newUploadBatchStats()
	for _, u := range uploads {
		read := sync.OnceValues(func() ([]byte, error) { return os.ReadFile(u.src) })
		for _, to := range u.to {
			g.Go(func() error {
				defer takeUploadSlot()()
				if u.replace {
					data, err := read()
					if err != nil {
						readErr := fmt.Errorf("read %s: %w", u.src, err)
						now := time.Now()
						stats.record(uploadOutcome{Start: now, End: now, Failed: true, Err: readErr})
						return readErr
					}
					return tracedPut(ctx, to.up, to.bucket, u.key, u.headers, data, stats)
				}
				return tracedUpload(ctx, to.up, to.bucket, u.key, u.headers, read, stats)
			})
		}
	}
	err := g.Wait()
	emitUploadBatch(report, uploadKindStaticAsset, stats, err, phaseStart)
	return err
}

File: platform/edge/cloudflare/workers/entry/src/assets.ts
const IMMUTABLE_CACHE_CONTROL = "public, max-age=31536000, immutable";
const REVALIDATE_CACHE_CONTROL = "public, max-age=0, must-revalidate";

export interface AssetObject {
  body: ReadableStream | null;
  httpEtag?: string;
}

export interface AssetBucket {
  get(key: string): Promise<AssetObject | null>;
}

export interface AssetStoreDeps {
  store?: AssetBucket;
  assetPrefix: string;
  basePath?: string;
  cache: Pick<Cache, "match" | "put">;
  waitUntil: (promise: Promise<unknown>) => void;
}

function table(entries: Record<string, string>): Map<string, string> {
  return new Map(Object.entries(entries));
}

const CONTENT_TYPES = table({
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".mjs": "text/javascript; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".json": "application/json; charset=utf-8",
  ".map": "application/json; charset=utf-8",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".avif": "image/avif",
  ".ico": "image/x-icon",
  ".woff": "font/woff",
  ".woff2": "font/woff2",
  ".ttf": "font/ttf",
  ".eot": "application/vnd.ms-fontobject",
  ".txt": "text/plain; charset=utf-8",
  ".xml": "application/xml",
  ".webmanifest": "application/manifest+json",
  ".wasm": "application/wasm",
});

const METADATA_CONTENT_TYPES = table({
  "robots.txt": "text/plain",
  "manifest.json": "application/manifest+json",
});

export function contentTypeFor(pathname: string): string {
  const name = pathname.slice(pathname.lastIndexOf("/") + 1).toLowerCase();
  const metadata = METADATA_CONTENT_TYPES.get(name);
  if (metadata) return metadata;
  const dot = name.lastIndexOf(".");
  if (dot === -1) return "application/octet-stream";
  return CONTENT_TYPES.get(name.slice(dot)) ?? "application/octet-stream";
}

const NEXT_STATIC_PREFIX = "/_next/static/";
const SERVICE_WORKER_PREFIX = "service-worker/";

export function isNextStaticPathname(pathname: string): boolean {
  return pathname.includes(NEXT_STATIC_PREFIX);
}

export function cacheControlFor(pathname: string): string {
  const at = pathname.indexOf(NEXT_STATIC_PREFIX);
  if (at === -1) return REVALIDATE_CACHE_CONTROL;
  const itemPath = pathname.slice(at + NEXT_STATIC_PREFIX.length);
  return itemPath.startsWith(SERVICE_WORKER_PREFIX)
    ? REVALIDATE_CACHE_CONTROL
    : IMMUTABLE_CACHE_CONTROL;
}

function storedPathnames(pathname: string, basePath = ""): string[] {
  if (pathname.endsWith(".html")) return [pathname];
  if (pathname === "/" || pathname === basePath) {
    return [
      pathname === "/" ? "/index.html" : `${pathname}/index.html`,
      pathname,
    ];
  }
  const document = `${pathname}.html`;
  return hasExtension(pathname) ? [pathname, document] : [document, pathname];
}

function hasExtension(pathname: string): boolean {
  return pathname.slice(pathname.lastIndexOf("/") + 1).includes(".");
}

async function notFound(deps: AssetStoreDeps, locale?: string): Promise<Response> {
  const names = locale ? [`/${locale}/404.html`, "/404.html"] : ["/404.html"];
  for (const name of names) {
    const page = await deps.store?.get(`${deps.assetPrefix}${name}`);
    if (page?.body) {
      return new Response(page.body, {
        status: 404,
        headers: { "content-type": "text/html; charset=utf-8" },
      });
    }
  }
  return new Response("Not Found", { status: 404 });
}

export async function serveStaticAsset(
  request: Request,
  url: URL,
  deps: AssetStoreDeps,
  locale?: string,
): Promise<Response> {
  if (!deps.store) return notFound(deps, locale);

  const cached = await deps.cache.match(request);
  if (cached) return cached;

  let hit:
    | { pathname: string; object: AssetObject & { body: ReadableStream } }
    | undefined;
  for (const pathname of storedPathnames(url.pathname, deps.basePath)) {
    const object = await deps.store.get(`${deps.assetPrefix}${pathname}`);
    if (object?.body) {
      hit = { pathname, object: { ...object, body: object.body } };
      break;
    }
  }
  if (!hit) return notFound(deps, locale);

  const { object } = hit;
  const cacheControl = cacheControlFor(hit.pathname);
  const headers = new Headers({
    "content-type": contentTypeFor(hit.pathname),
    "cache-control": cacheControl,
  });
  if (object.httpEtag) headers.set("etag", object.httpEtag);

  if (object.httpEtag && matchesEtag(request, object.httpEtag)) {
    await object.body.cancel();
    return new Response(null, { status: 304, headers });
  }

  const response = new Response(object.body, { status: 200, headers });
  if (cacheControl === IMMUTABLE_CACHE_CONTROL) {
    deps.waitUntil(deps.cache.put(request, response.clone()));
  }
  return response;
}

function matchesEtag(request: Request, etag: string): boolean {
  const header = request.headers.get("if-none-match");
  if (!header) return false;
  if (header.trim() === "*") return true;
  const strong = (tag: string) => tag.trim().replace(/^W\//, "");
  return header.split(",").some((tag) => strong(tag) === strong(etag));
}
